runners/core-java/src/test/java/org/apache/beam/runners/core/SimpleDoFnRunnerTest.java

```diff
@@ -198,44 +218,185 @@ public void testFinishBundleExceptionsWrappedAsUserCodeException() {
+                DoFnWithTimers.PROCESSING_TIMER_ID,
                 StateNamespaces.window(windowFn.windowCoder(), GlobalWindow.INSTANCE),
-                currentTime.plus(offset),
-                TimeDomain.EVENT_TIME)));
+                currentProcessingTime.plus(DoFnWithTimers.TIMER_OFFSET).plus(1),
+                TimeDomain.PROCESSING_TIME),
+            TimerData.of(
+                DoFnWithTimers.EVENT_TIMER_ID,
+                StateNamespaces.window(windowFn.windowCoder(), GlobalWindow.INSTANCE),
+                currentEventTime.plus(DoFnWithTimers.TIMER_OFFSET),
+                TimeDomain.EVENT_TIME));
+
+    pipeline.run();
+  }
+
+  /**
+   * Tests that {@link SimpleDoFnRunner#onTimer} properly dispatches to the underlying {@link DoFn}
+   * on appropriate time domains. With {@link IntervalWindow}, we check behavior of emitted events
+   * when time is inside and outside of window boundaries.
+   */
+  @Test
+  @Category(NeedsRunner.class)
+  public void testOnTimerCalledWithIntervalWindow() {
+
+    // TIMESTAMP_MIN_VALUE is initial value for processing time used done by TestClock
+    Instant baseTime = new Instant(0);
+
+    Duration windowDuration = Duration.standardHours(1);
+    Duration windowLateness = Duration.standardMinutes(1);
+    IntervalWindow window = new IntervalWindow(baseTime, windowDuration);
+    FixedWindows windowFn = FixedWindows.of(windowDuration);
+    DoFnWithTimers<IntervalWindow> fn = new DoFnWithTimers<>(windowFn.windowCoder());
+
+    TimestampedValue<KV<String, String>> event =
+        TimestampedValue.of(KV.of("anyKey", "anyValue"), window.start());
+
+    TestStream<KV<String, String>> testStream =
+        TestStream.create(KvCoder.of(StringUtf8Coder.of(), StringUtf8Coder.of()))
+            // watermark in window, processing time far behind
+            .advanceWatermarkTo(window.start())
+            .addElements(event)
+            .advanceProcessingTime(DoFnWithTimers.TIMER_OFFSET.plus(1))
+            .advanceWatermarkTo(window.start().plus(DoFnWithTimers.TIMER_OFFSET).plus(1))
+
+            // watermark and processing time within window
+            .advanceProcessingTime(
+                Duration.millis(Math.abs(BoundedWindow.TIMESTAMP_MIN_VALUE.getMillis())))
+            .addElements(event)
+            .advanceProcessingTime(DoFnWithTimers.TIMER_OFFSET.plus(1))
+            .advanceWatermarkTo(
+                window.start().plus(2 * DoFnWithTimers.TIMER_OFFSET.getMillis()).plus(2))
+
+            // watermark in window, processing time ahead of window.end() but within lateness
+            .advanceProcessingTime(windowDuration)
+            .addElements(event)
+            .advanceProcessingTime(DoFnWithTimers.TIMER_OFFSET.plus(1))
+            .advanceWatermarkTo(
+                window.start().plus(3 * DoFnWithTimers.TIMER_OFFSET.getMillis()).plus(3))
+
+            // watermark in window, processing time  is out of window's allowed lateness
```

Comment:
If this doesn't cause a crash, then we have a bug!

Reply:
So, what now. I have no idea where to look (did not try yet to be frank) :) Which part decides on valid / dropped elements?

Reply:
Hey @kennknowles , any comments? Should I add dropping at SimpleDoFnRunner?

Reply:
I think a more isolated test case might be easier. You want processing time to be way beyond the allowed lateness. Then when you emit elements I think `outputWithTimestamp` should actually crash saying that the timestamp was out of bounds.

Reply:
What kind of isolation do you have in mind? I was under the impression that if one wants to processElements, set timer and get timer callback called, test pipeline is needed. Is there example I can follow?

I am still unclear on what needs to be dropped when by who. Dynamics of DoFnRunner is quite complex. Test case does not use `outputWithTimestamp` but only `output`. Both just constructs  `WindowedValue` and fire them away.

Also, there are no limitations on for when timer can be set, but there are some event time.
https://github.com/apache/beam/blob/master/runners/core-java/src/main/java/org/apache/beam/runners/core/SimpleDoFnRunner.java#L995

Reply:
Without further input, I can not finish this. Sorry